mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -63,6 +64,48 @@ class Completions(
     case (_: Ident) :: (_: SeqLiteral) :: _ => false
     case _ => true
 
+  private def calculateTypeInstanceAndNewPositions() =
+
+    path match
+      case head :: tail
+          if head.isInstanceOf[Select] || head.isInstanceOf[Ident] =>
+        shouldAddSquareBracket =
+          val span: Span = head.srcPos.span
+          if span.exists then
+            var i = span.end
+            while i < (text.length() - 1) && text(i).isWhitespace do i = i + 1
+
+            if (i < text.length()) then text(i) != '['
+            else true
+          else false
+        tail match
+          case (v: ValOrDefDef) :: _ =>
+            if v.tpt.sourcePos.contains(pos) then isTypePosition = true
+          case New(selectOrIdent) :: _
+              if selectOrIdent.isInstanceOf[Select] || selectOrIdent
+                .isInstanceOf[Ident] =>
+            if selectOrIdent.sourcePos.contains(pos) then isNewPosition = true
+          case (a @ AppliedTypeTree(_, args)) :: _ =>
+            if args.exists(_.sourcePos.contains(pos)) then isTypePosition = true
+            else shouldAddSquareBracket = false
+          case (_: Import) :: _ =>
+          case _ =>
+            isInstantiationOrMethodCallPos = true
+        end match
+
+      case (_: TypeTree) :: TypeApply(Select(newQualifier: New, _), _) :: _
+          if newQualifier.sourcePos.contains(pos) =>
+        isNewPosition = true
+
+      case _ =>
+    end match
+  end calculateTypeInstanceAndNewPositions
+
+  private var isTypePosition = false
+  private var isNewPosition = false
+  private var isInstantiationOrMethodCallPos = false
+  private var shouldAddSquareBracket = true
```

Comment:
Can we make all of them return values of `calculateTypeInstanceAndNewPositions`? (making them immutable?)
we shouldn't use mutable fields as much as possible if it's not needed because it makes it hard to debug (in the future) / understand the code in general.

Reply:
Good idea 👍 I would try and see. I remember once when I was calling `calculateTypeInstanceAndNewPositions` for doing that though, and surprisingly, the values were not calculated correctly. I need to see this time.